### PR TITLE
Add framework for managing segment id pools.

### PR DESCRIPTION
**Pull request by morgabra**
Add API backed segment_id range management.

Also added first use - allocating vxlan VNIs in the NVP driver.

Here's a sample session of the whole thing working:

```
# add a segment allocation range
$ curl localhost:9696/v2.0/segment_allocation_ranges -X POST -H 'Content-Type: application/json' -d 

# and show it
$ curl localhost:9696/v2.0/segment_allocation_ranges/3c800bcc-b198-4f1b-87b1-329a0b806056 | python -mjson.tool
{
    "segment_allocation_range": {
        "do_not_use": false,
        "first_id": 1000,
        "free_ids": 6,
        "id": "3c800bcc-b198-4f1b-87b1-329a0b806056",
        "last_id": 1005,
        "segment_id": "3e2b4fcd-6875-44fe-acae-e640b625a216",
        "segment_type": "vxlan",
        "size": 6
    }
}

# notice 'segment_id' in this case is the tz id, the nvp driver knows which segments to use
# add a network
$ neutron net-create morgabra-test --tenant_id admin
Created a new network:
+----------------+--------------------------------------+
| Field          | Value                                |
+----------------+--------------------------------------+
| admin_state_up | True                                 |
| id             | e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce |
| name           | morgabra-test                        |
| shared         | False                                |
| status         | ACTIVE                               |
| subnets        |                                      |
| tenant_id      | admin                                |
+----------------+--------------------------------------+

DEBUG quark.drivers.nvp_driver [req-cf406eda-67f3-41cf-b011-f3d9f0377435 - admin - - -] Creating new lswitch for admin network morgabra-test _lswitch_create /Users/morgabra/work/src/teeth/quark/quark/drivers/nvp_driver.py:612
DEBUG quark.drivers.nvp_driver [req-cf406eda-67f3-41cf-b011-f3d9f0377435 - admin - - -] Creating lswitch for network e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce _lswitch_create /Users/morgabra/work/src/teeth/quark/quark/drivers/nvp_driver.py:625
INFO quark.segment_allocations [req-cf406eda-67f3-41cf-b011-f3d9f0377435 - admin - - -] Attempting to allocate segment for network e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce segment_id 3e2b4fcd-6875-44fe-acae-e640b625a216 segment_type vxlan
INFO quark.segment_allocations [req-cf406eda-67f3-41cf-b011-f3d9f0377435 - admin - - -] Allocated segment 1000 for network e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce segment_id 3e2b4fcd-6875-44fe-acae-e640b625a216 segment_type vxlan
INFO aiclib.core [req-cf406eda-67f3-41cf-b011-f3d9f0377435 - admin - - -] (POST @ /ws.v1/lswitch): {'transport_zones': [{'zone_uuid': '3e2b4fcd-6875-44fe-acae-e640b625a216', 'transport_type': 'stt'}, {'zone_uuid': '3e2b4fcd-6875-44fe-acae-e640b625a216', 'binding_config': {'vxlan_transport': [{'transport': 1000L}]}, 'transport_type': 'vxlan'}], 'display_name': u'morgabra-test', 'tags': [{'scope': 'os_tid', 'tag': u'admin'}, {'scope': 'neutron_net_id', 'tag': 'e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce'}]}

# seemed to have allocated a vni for the vxlan binding, and it's currently on the lswitch as expected
# check the range again
$ curl localhost:9696/v2.0/segment_allocation_ranges/3c800bcc-b198-4f1b-87b1-329a0b806056 | python -mjson.tool
{
    "segment_allocation_range": {
        "do_not_use": false,
        "first_id": 1000,
        "free_ids": 5,
        "id": "3c800bcc-b198-4f1b-87b1-329a0b806056",
        "last_id": 1005,
        "segment_id": "3e2b4fcd-6875-44fe-acae-e640b625a216",
        "segment_type": "vxlan",
        "size": 6
    }
}

# and of course it's freed when deleted
$ neutron net-delete e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce
Deleted network: e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce

INFO aiclib.nvp [req-6b01f29d-e23b-46e9-b3cb-ebf3c66b32fc - - - - -] (DELETE @ /ws.v1/lswitch/6a34b096-aa09-4ace-a4d8-eea71789e06f):{}:Elapsed time 0.143
INFO quark.segment_allocations [req-6b01f29d-e23b-46e9-b3cb-ebf3c66b32fc - - - - -] Attempting to deallocate segment for network e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce segment_id 3e2b4fcd-6875-44fe-acae-e640b625a216 segment_type vxlan
INFO quark.segment_allocations [req-6b01f29d-e23b-46e9-b3cb-ebf3c66b32fc - - - - -] Deallocated 1 allocated segment(s) for network e4179ec3-749b-4c2f-821b-bd8d4b8ea1ce segment_id 3e2b4fcd-6875-44fe-acae-e640b625a216 segment_type vxlan

curl localhost:9696/v2.0/segment_allocation_ranges/3c800bcc-b198-4f1b-87b1-329a0b806056 | python -mjson.tool
{
    "segment_allocation_range": {
        "do_not_use": false,
        "first_id": 1000,
        "free_ids": 6,
        "id": "3c800bcc-b198-4f1b-87b1-329a0b806056",
        "last_id": 1005,
        "segment_id": "3e2b4fcd-6875-44fe-acae-e640b625a216",
        "segment_type": "vxlan",
        "size": 6
    }
}

# and for fun
02:03 $ time curl localhost:9696/v2.0/segment_allocation_ranges -X POST -H 'Content-Type: application/json' -d '{"segment_allocation_range": {"segment_id": "3e2b4fcd-6875-44fe-acae-e640b625a216", "segment_type": "vxlan", "first_id": 2000000, "last_id": 3000000, "tenant_id": "admin"}}'
real	0m41.597s

$ curl localhost:9696/v2.0/segment_allocation_ranges/4759888c-d847-4993-bd26-ac11b5b75d77 | python -mjson.tool
{
    "segment_allocation_range": {
        "do_not_use": false,
        "first_id": 2000000,
        "free_ids": 1000001,
        "id": "4759888c-d847-4993-bd26-ac11b5b75d77",
        "last_id": 3000000,
        "segment_id": "3e2b4fcd-6875-44fe-acae-e640b625a216",
        "segment_type": "vxlan",
        "size": 1000001
    }
}

```
.

### CODE REVIEW

File: quark/db/models.py [hunk 1]
```diff
@@ -558,3 +558,43 @@ class LockHolder(BASEV2):
...
+
+
+class SegmentAllocation(BASEV2):
+    """A segment allocation."""
+    __tablename__ = "quark_segment_allocations"
+
+    # a particular segment id is unique across the segment and type - this data
+    # is denormalized to give us some safety around allocations, as well
+    # as allow us to look up allocations to reallocate without a join.
+    id = sa.Column(sa.BigInteger(), primary_key=True, autoincrement=False)
```
Comment (wmridley) [inline quark/db/models.py]: This is the VNI, right?

**Comment by roaet** [comment]
Super clean. I like it.

**Comment by morgabra** [comment]
Added tests and whatnot.

I'm really not very happy with where I shoehorned this into the NVP driver. I'm a little confused about how provider net attributes work (are they even a thing in quark anymore?) and the concept of 'default transport zones' in general.

I think the whole thing could use a rewrite, I tried to abstract the actual transport zone allocation details away from the driver a bit but it might be more work than it's worth.

**Comment by morgabra** [comment]
This is ready for a serious review now I think, and I'd love some guidance/suggestions around where to actually do the allocation at lswitch create() time.

**Comment by roaet** [comment]
+1

**Comment by mpath** [comment]
:+1:

**Comment by CrashenX** [comment]
Amazing team work here, thanks! We'll work on testing a package in preprod2-iad tomorrow.

**Pull request merged**